torchdata/datapipes/iter/load/s3io.py

```diff
@@ -0,0 +1,107 @@
+from io import BytesIO
+from typing import Iterator, Tuple
+
+import torchdata
+from torchdata.datapipes import functional_datapipe
+from torchdata.datapipes.iter import IterDataPipe
+from torchdata.datapipes.utils import StreamWrapper
+
+
+@functional_datapipe("list_file_by_s3")
+class S3FileListerIterDataPipe(IterDataPipe[str]):
+    r""":class:`S3FileListerIterDataPipe`.
+
+    Iterable DataPipe that lists URLs with the given prefixes (functional name: ``list_file_by_s3``).
+
+    Args:
+        source_datapipe: a DataPipe that contains URLs/URL prefixes to s3 files
+        length: Nominal length of the datapipe
+        requestTimeoutMs: optional, overwrite the default timeout setting for this datapipe
+        region: optional, overwrite the default region inferred from credentials for this datapipe
+
+    Note:
+        AWS_CPP_SDK is necessary to use the S3 DataPipe(s).
+
+    Example:
+        >>> from torchdata.datapipes.iter import S3FileLister, S3FileLoader
+        >>> s3_prefixes = ['s3://bucket-name/folder/', ...]
+        >>> dp_s3_urls = S3FileLister(s3_prefixes)
+        >>> dp_s3_files = S3FileLoader(s3_urls) # outputs in (url, StreamWrapper(BytesIO))
+        >>> # more datapipes to convert loaded bytes, e.g.
+        >>> datapipe = dp_s3_files.parse_csv(delimiter=' ')
+        >>> for d in datapipe: # Start loading data
+        ...     pass
+    """
+
+    def __init__(self, source_datapipe: IterDataPipe[str], length: int = -1, request_timeout_ms=-1, region="") -> None:
+        if not hasattr(torchdata, "_torchdata") or not hasattr(torchdata._torchdata, "S3Handler"):
+            raise ModuleNotFoundError("Torchdata must be built with BUILD_S3=1 to use this datapipe.")
+
+        self.source_datapipe: IterDataPipe[str] = source_datapipe
+        self.length: int = length
+        self.handler = torchdata._torchdata.S3Handler(request_timeout_ms, region)
+
+    def __iter__(self) -> Iterator[str]:
+        for prefix in self.source_datapipe:
+            while True:
+                urls = self.handler.list_files(prefix)
+                yield from urls
+                if not urls:
+                    break
```

Comment:
Please correct me if I am wrong. I am not sure how this `while` loop exits. When `urls` becomes None?
Does that mean `handler` understand what `urls` have been yielded?

Reply:
Yes, when `urls` becomes None, the loop exits. `handler` outputs at most 1000 items each time, and has a marker on where to continue in the next loop, until all items given the prefix are yieled.

Reply:
I see. It makes sense, thanks for explaining.